agent/submatview/store.go

```diff
@@ -0,0 +1,222 @@
+	for {
+		s.lock.RLock()
+		timer := s.expiryHeap.Next()
+		s.lock.RUnlock()
+
+		select {
+		case <-ctx.Done():
+			timer.Stop()
+			return
+		case <-s.expiryHeap.NotifyCh:
+			timer.Stop()
+			continue
+
+		case <-timer.Wait():
+			s.lock.Lock()
+
+			he := timer.Entry
+			s.expiryHeap.Remove(he.Index())
+
+			e := s.byKey[he.Key()]
+
+			// Only stop the materializer if there are no active requests.
+			if e.requests == 0 {
+				e.stop()
+				delete(s.byKey, he.Key())
+			}
+
+			s.lock.Unlock()
+		}
+	}
+}
+
+// TODO: godoc
+type Request interface {
+	cache.Request
+	NewMaterializer() (*Materializer, error)
+	Type() string
+}
+
+// Get a value from the store, blocking if the store has not yet seen the
+// req.Index value.
+// See agent/cache.Cache.Get for complete documentation.
+func (s *Store) Get(ctx context.Context, req Request) (Result, error) {
+	info := req.CacheInfo()
+	key, e, err := s.getEntry(req)
+	if err != nil {
+		return Result{}, err
+	}
+	defer s.releaseEntry(key)
+
+	ctx, cancel := context.WithTimeout(ctx, info.Timeout)
+	defer cancel()
+
+	result, err := e.materializer.getFromView(ctx, info.MinIndex)
+	// context.DeadlineExceeded is translated to nil to match the behaviour of
+	// agent/cache.Cache.Get.
+	if err == nil || errors.Is(err, context.DeadlineExceeded) {
+		return result, nil
+	}
+	return result, err
+}
+
+// Notify the updateCh when there are updates to the entry identified by req.
+// See agent/cache.Cache.Notify for complete documentation.
+//
+// Request.CacheInfo().Timeout is ignored because it is not really relevant in
+// this case. Instead set a deadline on the context.
+func (s *Store) Notify(
+	ctx context.Context,
+	req Request,
+	correlationID string,
+	updateCh chan<- cache.UpdateEvent,
+) error {
+	info := req.CacheInfo()
+	key, e, err := s.getEntry(req)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		defer s.releaseEntry(key)
+
+		index := info.MinIndex
+		for {
+			result, err := e.materializer.getFromView(ctx, index)
+			switch {
+			case ctx.Err() != nil:
+				return
+			case err != nil:
+				// TODO: cache.Notify sends errors on updateCh, should this do the same?
+				// It seems like only fetch errors would ever get sent along and eventually
+				// logged, so sending may not provide any benefit here.
+
+				s.logger.Warn("handling error in Store.Notify",
+					"error", err,
+					"request-type", req.Type(),
+					"index", index)
+				continue
+			}
+
+			index = result.Index
+			u := cache.UpdateEvent{
+				CorrelationID: correlationID,
+				Result:        result.Value,
+				Meta:          cache.ResultMeta{Index: result.Index},
+				Err:           err,
+			}
+			select {
+			case updateCh <- u:
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+	return nil
+}
+
+// getEntry from the store, and increment the requests counter. releaseEntry
+// must be called when the request is finished to decrement the counter.
+func (s *Store) getEntry(req Request) (string, entry, error) {
```

Comment:
I think this should return a materializer instead of an `entry` as return-value-2. The only field off of the entry we use is that one. Other fields on `entry` are misleading (`requests` (a copy)) or unsafe (`stop`) for callers to get access to.

Reply:
Ah yes, true! That looks like a good change